### ISSUE TITLE
Need "stop and let ring"

### ISSUE THREAD

**Issue by Learjeff** [powertab/powertabeditor#278]
As implemented, the "let ring" function isn't useful for its most common usage, which is like depressing the sustain pedal on a piano, playing a chord or arpeggio, and then lifting and immediately depressing the sustain pedal at the beginning of a new chord or arpeggio.

As it's currently implemented, you can't do this with "let ring," you have to use ties (which is tedious!)

Instead, we need two events: "start ringing" and "continue ringing."  With "start ringing," all notes would be triggered, regardless of whether the previous note or chord was letting ring.  "Continue ringing" would be implemented the same as current "let ring."

The tool buttons should be "let ring" and "keep ringing."

When applying the "let ring" attribute to a selected range of chords/notes, the first one selected should be inserted as "start ringing" and the remainder should be inserted as "continue ringing."

For notation, "start ringing" would force a vertical bar showing a discontinuity in ringing, above the tab, even if the previous note or chord was ringing.  Something like this:

```
     | let ring -----| let ring ----
E|-----------2---------------3-------|
B|-------3---------------0-----------|
G|---2---------------0---------------|
```
When interpreting existing tabs, the current "let ring" should be interpreted the same as before, as my suggested "continue ringing."  This way, existing tabs would not be affected.

BTW with some help getting started I could probably code this.

I suspect this is what [BUG: Let Ring Markers Not Implemented Correctly](http://www.power-tab.net/getarticle.php?id=84) was really about.

**Comment by cameronwhite** [comment]
If I'm understanding the link correctly, it's just a rendering issue where the marker isn't drawn properly?

This code would be in `createConnectedSymbolGroup()` in `painters/systemrenderer.cpp`.

**Comment by Learjeff** [comment]
The issue is both rendering and playback.  It is not possible to have a chord ring up to a new chord and have the old chord stop just as the new chord is played (using "let ring").  Yet that is the most common use case.

**Comment by cameronwhite** [comment]
Right, that makes sense now that I've actually tried it out :)
So yeah, I think having an "end let ring" marker or something along those lines would work. You could look at how bends are implemented for some inspiration, since they have a more complex system of start / hold / end events.

**Comment by Learjeff** [comment]
Sure, and "end let ring" would work too.  It'll be a while before I can get to this, in case it's an easy fix for anyone familiar with the code.  I will post about any progress I make.
Thanks for your attention on this.

**Comment by jlearman** [comment]
Hi Cameron.  I'm trying to get set up to work on this, but am stymied at the CMake step.  BTW, great README, very helpful.  But of course there are always too many variables to cover every situation.
Normally I use cygwin, but I'm not sure how to get CMake to cooperate with it, even though cygwin64/bin is in my %PATH%, and typical Linux programs like "ls" work from Windows csh.
I've searched for a solution but none seem to apply.  If you know a quick fix I'd appreciate that.  But perhaps it would be better to just use whatever you or most people use (possibly VisualC++, assuming there's still a free version.)  Also I've used Eclipse in the past and could try to set that up and perhaps CMake will then be able to generate makefiles for Eclipse.
Any suggestions?
BTW I'm an embedded real-time and networking product software developer for 40 years, doing mostly cross-development on Linux.  I don't have experience on open source but I'd like to get started.  If there's a better place for this discussion please point me, or feel free to reply by email.
Thanks!

**Comment by cameronwhite** [comment]
I'd probably stick to using the free version of MSVC (which works great), or using Eclipse. I'm not sure if I can help too much with any Cygwin issues :)

**Comment by jlearman** [comment]
Thanks!  I've installed MSVC and am up to the CMAKE step.  Should I use the Windows CMake application, or use CMake in MS Visual Studio?

Using the one in MSVS, I'm able to add the variables in CMakeSettings.json, but am unsure on how to do the "Configure" step.  "Project -> Generate cache" runs and generates warnings about boost version (oops I'm using a later version) but no errors.  Yet nothing under the Build menu gets activated, they're still grayed out.

**Comment by jlearman** [comment]
Hmm.  I do see this, which looks like an error exit:
cmd.exe /c ""C:\PROGRAM FILES (X86)\MICROSOFT VISUAL STUDIO\2019\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\CMake\bin\cmake.exe"  -G "Ninja" -DCMAKE_INSTALL_PREFIX:PATH="C:/Users/jlearman/jeff/Source/Repos/out/install/x64-Debug" -DBoost_INCLUDE_DIR:PATH="C:/Program Files/boost/boost_1_56_0" -DCMAKE_BUILD_TYPE:STRING="Debug" -DCMAKE_INSTALL_PREFIX:PATH="C:/Users/jlearman/jeff/Source/Repos/out/install/x64-Debug" -DCMAKE_MAKE_PROGRAM:FILEPATH="C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/Ninja/ninja.exe" -DCMAKE_OSX_DEPLOYMENT_TARGET:STRING="10.9" -DCMAKE_PREFIX_PATH:STRING="/Qt/5.14.1/msvc2015_64/lib/cmake" -DCMAKE_C_COMPILER:FILEPATH="C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.25.28610/bin/HostX64/x64/cl.exe" -DCMAKE_CXX_COMPILER:FILEPATH="C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.25.28610/bin/HostX64/x64/cl.exe"  -DCMAKE_BUILD_TYPE="Debug" "C:\Users\jlearman\jeff\Source\Repos" 2>&1" returned with exit code: 1
Yet all the messages from CMake are warnings.

**Comment by jlearman** [comment]
So I replaced boost with the version linked in README and didn't get as far:
1> [CMake] CMake Error at C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:146 (message):
1> [CMake]   Could NOT find Boost (missing: date_time filesystem iostreams
1> [CMake]   program_options regex system zlib) (found suitable version "1.56.0",
1> [CMake]   minimum required is "1.54.0")

I know it's not your job to sort this out but if you have any suggestions I'd appreciate it, thanks!

**Comment by cameronwhite** [comment]
Hey Jeff, I'm not sure about the first error without seeing the rest of the
log. Using either the built-in CMake or the normal Windows app should be
fine though
In the second error it sounds like it correctly found Boost 1.56 but didn't
find any of the compiled libraries. Maybe try setting BOOST_LIBRARYDIR if
those libraries do exist in your installation but aren't being picked up?

**Comment by jlearman** [comment]
Thanks, found my mistake.  (boost 1.56 didn't build successfully for whatever reason, and I had a typo that caused the error with the current version.)  I've successfully built in MSVS using "Build all".  But the Debug->Start is inactive, so how will I debug what I built?  I can run it from build/x64 Debug/bin and it works.  I know I can attach to it after running it, if necessary, but I bet there's a better way.

**Comment by cameronwhite** [comment]
Great! I think you need to change the startup project, which I think is in the right click menu in the project view.

**Comment by jlearman** [comment]
Thanks for the tip.  That's what all the web results say too, but "set startup" doesn't appear until I find the executable and right click on it (in the solution explorer).  Anyway, that works, great.  Next: this PTE is rather different from the one I was using.  First, the edit and playback windows are the same (great!)  With the one I'd been using, playing made the window change (toolbars disappeared, etc.) and it was hard to keep my place in the song.  Now it plays "in place" which is great.  However, there are no toolbars for notes etc, and toolbars don't appear anywhere in the menu.  Right clicking on the menu bar offers only mixer and instruments.  Perhaps it's built wrong?

**Comment by jlearman** [comment]
Ah I see I missed a lot of steps, which are written for using CMake app rather than MSVS.  Stay tuned, I'll try to sort it out.

**Comment by jlearman** [comment]
Aha -- those were PTE 1.7 features that aren't implemented in PTE 2.

**Comment by cameronwhite** [comment]
Yeah, there are a number of UI differences from 1.7, some are intended and others are just unimplemented so far
